File: tests/primitives/str_length/str_length.fs
module rec str_length.Program

// string has `Length` property.

let main _ =
  let hello = "hello"
  let world = "world"

  assert (hello.Length = 5)

  let message = hello + world
  assert (message.Length = 10)

  assert (hello.Length = 5)

  // FIXME: Hex escape other than \x00 is unimplemented.
  // assert ("\x00\x01\x02\x03\x04".Length = 5)
  assert ("\x00ABCD".Length = 5)

  // unicode character length
  assert ("あ".Length = 3)
  0

File: tests/primitives/str_length/str_length.c
#include <milone.h>

int main(int argc, char** argv);

int main(int argc, char** argv) {
    milone_start(argc, argv);
    struct String hello_ = (struct String){.str = "hello", .len = 5};
    struct String world_ = (struct String){.str = "world", .len = 5};
    milone_assert((hello_.len == 5), (struct String){.str = "str_length/str_length.milone", .len = 28}, 8, 2);
    struct String message_ = str_add(hello_, world_);
    milone_assert((message_.len == 10), (struct String){.str = "str_length/str_length.milone", .len = 28}, 11, 2);
    milone_assert((hello_.len == 5), (struct String){.str = "str_length/str_length.milone", .len = 28}, 13, 2);
    milone_assert((5 == 5), (struct String){.str = "str_length/str_length.milone", .len = 28}, 15, 2);
    milone_assert((3 == 3), (struct String){.str = "str_length/str_length.milone", .len = 28}, 18, 2);
    return 0;
}
